### PR TITLE
Added workspacesDesiredBundleType plugin and spec

### PR DESCRIPTION
**Pull request by m-nouman-suleman**
## Task
- https://scalock.atlassian.net/browse/SAAS-3110

### CODE REVIEW

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    more_info: 'A bundle in AWS WorkSpaces defines the hardware and software for AWS WorkSpaces. You can create a WorkSpaces instance using a predefined or custom bundle. Setting a limit to the types that can be used will help you control billing and address internal compliance requirements.',
+    recommended_action: 'Ensure that WorkSpaces instances are using desired bundle types',
+    link: 'https://docs.aws.amazon.com/workspaces/latest/adminguide/amazon-workspaces-bundles.html',
+    apis: ['WorkSpaces:describeWorkspaces'],
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+
```
Comment (AkhtarAmir) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:18]: Should have a setting for desired bundle types and compare each workspace against that type

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/workspaces/workspacesDesiredBundleType.js (hunk 1 above)
Comment (m-nouman-suleman) [reply plugins/aws/workspaces/workspacesDesiredBundleType.js:26]: Added the settings for desired bundle type but we have to  check if the combined workspaces in a specific region have a single bundle type and if not this plugin will fail
As per the cloud conformity doc referenced here in the ticket https://scalock.atlassian.net/browse/SAAS-3110

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    title: 'WorkSpaces Desired Bundle Type',
+    category: 'WorkSpaces',
+    domain: 'Identity Access and Management',
+    description: 'Ensure that AWS WorkSpaces bundles are of desired types.',
+    more_info: 'A bundle in AWS WorkSpaces defines the hardware and software for AWS WorkSpaces. You can create a WorkSpaces instance using a predefined or custom bundle. Setting a limit to the types that can be used will help you control billing and address internal compliance requirements.',
+    recommended_action: 'Ensure that WorkSpaces instances are using desired bundle types',
+    link: 'https://docs.aws.amazon.com/workspaces/latest/adminguide/amazon-workspaces-bundles.html',
+    apis: ['WorkSpaces:describeWorkspaces'],
+    settings: {
+        desired_bundle_type: {
```
Comment (m-nouman-suleman) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:14]: workspace desired bundle type

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            name: 'Workspaces desired bundle type.',
+            description: 'Plugin results will become aggregated once this value is breached',
```
Comment (m-nouman-suleman) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:16]: Comma separated list of desired bundle types

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            name: 'Workspaces desired bundle type',
+            description: 'Comma separated list of desired Workspace bundle types',
+            regex: '^.*$',
+            default: ''
+        }
+    },
+
+    run: function(cache, settings, callback) {
+        var workspace_desired_bundle_type = settings.workspace_desired_bundle_type || this.settings.workspace_desired_bundle_type.default;
+        if (!workspace_desired_bundle_type.length) return callback();
```
Comment (AkhtarAmir) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:24]:
```suggestion

        if (!workspace_desired_bundle_type.length) return callback(null, results, source);
```

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                helpers.addResult(
+                    results, 0, 'No Workspaces found.', region);
+                return rcb();
+            }
+
+            listWorkspaces.data.forEach(workspace => {
+                var resource = 'arn:' + awsOrGov + ':workspaces:' + region + ':' + accountId + ':workspace/' + workspace.WorkspaceId;
+
+                if (workspace.WorkspaceProperties && workspace.WorkspaceProperties.ComputeTypeName
+                    && workspace_desired_bundle_type.includes(workspace.WorkspaceProperties.ComputeTypeName.toUpperCase())) {
```
Comment (AkhtarAmir) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:57]:
```suggestion
                if (workspace.WorkspaceProperties && workspace.WorkspaceProperties.ComputeTypeName && workspace_desired_bundle_type.toUpperCase().includes(workspace.WorkspaceProperties.ComputeTypeName.toUpperCase())) {
```

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                    helpers.addResult(results, 0,
+                        'Workspace is using the desired bundle type', region, resource);
+                } else {
+                    helpers.addResult(results, 2,
+                        'Workspace is not using the desired bundle type', region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:62]:
```suggestion
                        'WorkSpaces instance is using the desired bundle type', region, resource);
                } else {
                    helpers.addResult(results, 2,
                        'WorkSpaces instance is not using the desired bundle type', region, resource);
```

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        var accountId = helpers.addSource(cache, source, ['sts', 'getCallerIdentity', acctRegion, 'data']);
+
+        async.each(regions.workspaces, function(region, rcb){
+            var listWorkspaces = helpers.addSource(cache, source, ['workspaces', 'describeWorkspaces', region]);
+
+            if (!listWorkspaces) {
+                return rcb();
+            }
+
+            if (listWorkspaces.err) {
```
Comment (AkhtarAmir) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:41]:
```suggestion
            if (listWorkspaces.err || !listWorkspaces.data) {
```

File: plugins/aws/workspaces/workspacesDesiredBundleType.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                helpers.addResult(
+                    results, 3, 'Unable to query for WorkSpaces information: ' + helpers.addError(listWorkspaces), region);
+                return rcb();
+            }
+
+            if (!listWorkspaces.data.length) {
+                helpers.addResult(
+                    results, 0, 'No Workspaces found.', region);
```
Comment (AkhtarAmir) [inline plugins/aws/workspaces/workspacesDesiredBundleType.js:49]:
```suggestion
                    results, 0, 'No WorkSpaces instances found', region);
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by dipsubha06 — APPROVED** [review]

**Pull request merged**